README
README

This project establishes an tree structure given a hierarchical coding
system. Once this tree structure is created it also computes the Euler
Circuit and corresponding levels in the tree [1]. The Euler Circuit
and the corresponding levels can be used to find lowest common
ancestors in the tree using range minimum queries. This application
was developed for the ICD-9 coding standard, but it is also
generalises to others.

The code is implemented in C++ using the boost libraries, in
particular the Boost Graph Library.


EXECUTION

The following command-line parameters are provided:

General Configuration:
  --help                produce help message
  --version             show the version

I/O Configuration:
  --result arg (=./results) results directory.
  --codes arg               CSV file with the codes.


The ICD-9 codes are expected to be in the following format for each
ICD-9 code:

Level-1, Level-2, ..., Level-n

Internally, a root vertex is created as a parent to all level-1
vertices in the input file. While the ICD-9 codes have balanced
levels, other coding systems might not. Internally, the names of the
vertices are prefixed with the current level. This is to ensure that
each vertex has a unique name.

[1] http://en.wikipedia.org/wiki/Eulerian_path


ChangeLog
* [0.5] - 2015-04-30
** Changed
 - Read the ICD-9 categories and compute the tree structure
 - Implement DFS to compute the Euler Circuit

* [0.6] - 2015-05-01
** Added
 - Serialise the Euler Circuit and corresponding levels
 - Serialise the ICD-9 tree structure into an Graph ML file

* [0.7] - 2015-05-05
** Fixed
 - Tree construction. Added a command-line argument to uniquify the vertex
	names using a level prefix.
** Added
 - Graphviz output


src/main/main.cc
// Copyright (C) 2015 Dominik Dahlem <Dominik.Dahlem@gmail.com>
//
// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program; if not, write to the Free Software
// Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.

#if HAVE_CONFIG_H
# include <config.h>
#endif

#ifndef __STDC_CONSTANT_MACROS
# define __STDC_CONSTANT_MACROS
#endif /* __STDC_CONSTANT_MACROS */

#include <fstream>
#include <iostream>
#include <vector>

#include <boost/cstdint.hpp>
#include <boost/tokenizer.hpp>

#include <boost/algorithm/string/split.hpp>
#include <boost/algorithm/string/trim.hpp>

#include <boost/graph/graphml.hpp>
#include <boost/graph/graphviz.hpp>
#include <boost/graph/properties.hpp>
#include <boost/graph/undirected_dfs.hpp>

#include "CL.hh"

#include "Graph.hh"
namespace igraph = graph;

#include "Types.hh"
namespace itypes = types;

typedef boost::tokenizer <boost::escaped_list_separator <char> > Tokenizer;


int main(int argc, char *argv[])
{
  args_t args;
  CL cl;

  if (cl.parse(argc, argv, args)) {
    return EXIT_SUCCESS;
  }

  std::ifstream codesFile;

  codesFile.open(args.codes.c_str());
  if (!codesFile.is_open()) {
    std::cerr << "Could not open file: " << args.codes << std::endl;
    return EXIT_FAILURE;
  }

#ifndef NDEBUG
  std::cout << "Reading codes..." << std::endl;
#endif /* NDEBUG */

  igraph::Graph g;

  // the ICD-9 tree is disjointed if we were not to introduce an artificial root vertex
  igraph::Vertex root = boost::add_vertex(g);
  g[root].level = 0;
  g[root].id = 1;
  g[root].name = "epsilon"; // root vertex name

  std::string line;
  while (!codesFile.eof()) {
    std::getline(codesFile, line);
#ifndef NDEBUG
    std::cout << "Read line: " << line << std::endl;
#endif /* NDEBUG */

    if (line != "") {
      itypes::StringVector cats;
      boost::split(cats, line, boost::is_any_of(","), boost::token_compress_on);

      // build the tree structure
      igraph::add(cats, g, root);
    }
  }
  codesFile.close();

  itypes::StringVector euler_circuit;
  itypes::IntVector levels;
  igraph::dfs_euler_circuit<itypes::StringVector, itypes::IntVector> dfs_euler(euler_circuit, levels);
  boost::depth_first_search(g, boost::visitor(dfs_euler));

  // serialise the ICD-9 tree, euler circuit and the corresponding levels
  std::string outEulerFile = args.results_dir + "/euler_circuit.dat";
  std::ofstream outEuler(outEulerFile.c_str(), std::ios::out);
  std::copy(euler_circuit.begin(), euler_circuit.end(), std::ostream_iterator<std::string>(outEuler, "\n"));
  outEuler.close();

  std::string outLevelsFile = args.results_dir + "/levels.dat";
  std::ofstream outLevels(outLevelsFile.c_str(), std::ios::out);
  std::copy(levels.begin(), levels.end(), std::ostream_iterator<boost::uint32_t>(outLevels, "\n"));
  outLevels.close();

  boost::dynamic_properties dp;
  dp.property("id", get(&igraph::VertexProperties::id, g));
  dp.property("Name", get(&igraph::VertexProperties::name, g));
  dp.property("Level", get(&igraph::VertexProperties::level, g));

  std::string outTreeFile = args.results_dir + "/tree.gml";
  std::ofstream outTree(outTreeFile.c_str(), std::ios::out);
  boost::write_graphml(outTree, g, get(&igraph::VertexProperties::id, g), dp, false);
  outTree.close();

  std::string outGVFile = args.results_dir + "/tree.dot";
  std::ofstream outGV(outGVFile.c_str(), std::ios::out);
  boost::write_graphviz(outGV, g, boost::make_label_writer(get(&igraph::VertexProperties::name, g)));
  outGV.close();

  return EXIT_SUCCESS;
}


src/main/CL.cc
// Copyright (C) 2015 Dominik Dahlem <Dominik.Dahlem@gmail.com>
//
// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program; if not, write to the Free Software
// Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.

/** @file CL.cc
 * Implementation of the command-line parsing of the main routine
 *
 * @author Dominik Dahlem
 */
#if HAVE_CONFIG_H
# include <config.h>
#endif

#ifndef __STDC_CONSTANT_MACROS
# define __STDC_CONSTANT_MACROS
#endif /* __STDC_CONSTANT_MACROS */

#include <iostream>

#include <boost/cstdint.hpp>
#include <boost/program_options/options_description.hpp>
#include <boost/program_options/parsers.hpp>
#include <boost/program_options/variables_map.hpp>
namespace po = boost::program_options;

#include "CL.hh"


CL::CL()
    : m_opt_desc(new po::options_description("Options"))
{
  // Declare the supported options.
  po::options_description opt_general("General Configuration");
  opt_general.add_options()
      (HELP.c_str(), "produce help message")
      (VERS.c_str(), "show the version")
      ;

  po::options_description opt_io("I/O Configuration");
  opt_io.add_options()
      (RESULTS_DIR.c_str(), po::value <std::string>()->default_value("./results"), "results directory.")
      (CODES.c_str(), po::value <std::string>()->default_value(""), "CSV file with the codes.")
      (PREFIX.c_str(), po::value <bool>()->default_value(0), "Add prefix to vertex names, if they are not unique.")
      ;

  m_opt_desc->add(opt_general);
  m_opt_desc->add(opt_io);
}


int CL::parse(int argc, char *argv[], args_t &p_args)
{
  po::variables_map vm;

  po::store(po::parse_command_line(argc, argv, (*m_opt_desc.get())), vm);
  po::notify(vm);

  if (vm.count(HELP)) {
    std::cout << (*m_opt_desc.get()) << std::endl;
    return EXIT_FAILURE;
  }

  if (vm.count(VERS)) {
    std::cout << PACKAGE_NAME << " " << PACKAGE_VERSION << std::endl;
    std::cout << argv[0] << std::endl;
    return EXIT_FAILURE;
  }

  if (vm.count(RESULTS_DIR.c_str())) {
    p_args.results_dir = vm[RESULTS_DIR.c_str()].as <std::string>();
  }

  if (vm.count(CODES.c_str())) {
    p_args.codes = vm[CODES.c_str()].as <std::string>();
  }

  if (vm.count(PREFIX.c_str())) {
    p_args.prefix = vm[PREFIX.c_str()].as <bool>();
  }

  std::cout << argv[0] << " " << PACKAGE_VERSION << std::endl;
  std::cout << PACKAGE_NAME << std::endl;
  std::cout << p_args << std::endl;

  return EXIT_SUCCESS;
}
